Add `grafana_folders` data source for fetching all folders

mdb opened the pull request:
This seeks to address issue #377 by adding a `grafana_folders` data source, which can be used to fetch all Grafana folders.

Note that this is an alternative implementation to that contained in PR #389, which I admittedly only noticed just now (I apologize) and which appears to have gone stale with git conflicts and an unsigned CLA. Would maintainers prefer I close this PR in favor of the PR #389 implementation?

mdb committed:
add basic `grafana_folders` data source

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
add grafana_folders data source tests

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
add `grafana_folders` to data source docs

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb commented on examples/data-sources/grafana_folders/data-source.tf:
```diff
@@ -0,0 +1,16 @@
+resource "grafana_folder" "test_b" {
+  title = "test-folder-b"
+  uid   = "test-ds-folder-uid-b"
+}
```
Would it be preferable if the example used a single `grafana_folder.test` and a `for_each`, rather than declaring semi-redundant `grafana_folder` resources?

julienduchesne replied:
All good like this

mdb commented on grafana/data_source_folders.go:
```diff
@@ -0,0 +1,95 @@
+	}
+	md5 := md5.Sum([]byte(strings.Join(uids, "-")))
+
+	d.SetId(fmt.Sprintf("%x", md5))
```
I believe an ID is required. In this case, the ID is an MD5 hash of a `-`-delimited string of all the folder UIDs. Is that acceptable?

mdb committed:
appease golangci-lint

This seeks to address:

```
grafana/data_source_folders.go:5:2: G501: Blocklisted import crypto/md5: weak cryptographic primitive (gosec)
  "crypto/md5"
  ^
grafana/data_source_folders.go:71:9: G401: Use of weak cryptographic primitive (gosec)
  md5 := md5.Sum([]byte(strings.Join(uids, "-")))
         ^
```

Signed-off-by: Mike Ball <mikedball@gmail.com>

julienduchesne commented on grafana/data_source_folders.go:
```diff
@@ -0,0 +1,95 @@
+	}
+	sha := sha256.Sum256([]byte(strings.Join(uids, "-")))
+
+	d.SetId(fmt.Sprintf("%x", sha))
```
The ID should be declarative because it needs to be unique. For a datasource like this, it's usually the combination of all entry parameters because you will typically not have the same datasource twice with the same parameters, it's useless

With no parameters, you can just set the ID as `grafana_folders`. No need to calculate a SHA

julienduchesne reviewed (commented):
Nice one. Just a small comment about the ID

mdb committed:
set grafana_folders ID to 'grafana_folders'

Per @julienduchesne (https://github.com/grafana/terraform-provider-grafana/pull/583/files#r944357467):

> The ID should be declarative because it needs to be unique. For a datasource like this, it's usually the combination of all entry parameters because you will typically not have the same datasource twice with the same parameters, it's useless
> With no parameters, you can just set the ID as grafana_folders. No need to calculate a SHA

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb commented:
> Nice one. Just a small comment about the ID

Thanks @julienduchesne - I've implemented your suggestion, assuming I interpreted correctly :)

julienduchesne reviewed (approved):
LGTM! Thanks!
